File: ISDESPOT/smart-car-sim-master/is-despot/problems/hybridVisual_car/src/SinCosLookupTable.cpp
#define _USE_MATH_DEFINES
#include <cmath>
#include <iostream>

class SinCosLookupTable {
    const double discretization = 0.005;

    double* cos_table = new double[(int) ceil((2 * M_PI) / discretization)];
    double* sin_table = new double[(int) ceil((2 * M_PI) / discretization)];

public:
    SinCosLookupTable() {
        for (int i = 0; i*discretization < 2 * M_PI; i++) {
            cos_table[i] = std::cos(i*discretization);
            sin_table[i] = std::sin(i*discretization);
        }
    }
    ~SinCosLookupTable(){
        delete[] cos_table;
        delete[] sin_table;
    }

    double sin(double radAngle) {
        if(radAngle < 0){
            radAngle += 2*M_PI;
        }

        return sin_table[(int) (fmod(radAngle, 2* M_PI) / discretization)];
    }

    double cos(double radAngle) {
        if(radAngle < 0){
            radAngle += 2*M_PI;
        }

        return cos_table[(int) (fmod(radAngle, 2* M_PI) / discretization)];
    }
};

File: tools/settingscontroller_client/src/Util/SinCosLookupTable.java
package settingscontroller_client.src.Util;

/**
 * A lookup table for sin and cos functions
 * for faster evaluation
 */
public class SinCosLookupTable {

    /**
     * Discretization of values to use
     */
    private double discretization = 0.001;

    /**
     * Stored cos values
     */
    private float[] cos = new float[(int) Math.ceil((2 * Math.PI) / discretization)];

    /**
     * Stored sin values
     */
    private float[] sin = new float[(int) Math.ceil((2 * Math.PI) / discretization)];

    /**
     * Static instance that can be used by all threads
     * asynchronously with ONLY reads
     */
    public static SinCosLookupTable table = new SinCosLookupTable();

    /**
     * Creates a lookup table with a discreatization between
     * values of discreatization
     */
    public SinCosLookupTable() {
        for (int i = 0; i*discretization < 2 * Math.PI; i++) {
            cos[i] = (float) Math.cos(i*discretization);
            sin[i] = (float) Math.sin(i*discretization);
        }
    }

    /**
     * Looks up sin(radAngle)
     * @param radAngle
     * @return sin(radAngle)
     */
    public float sin(double radAngle) {
        return sin[(int) (radAngle / discretization)];
    }

    /**
     * Looks up cos(radAngle)
     * @param radAngle
     * @return cos(radAngle)
     */
    public float cos(double radAngle) {
        return cos[(int) (radAngle / discretization)];
    }

}
